### ISSUE TITLE
Download failing after 96%  , 99% ( No network issues)

### ISSUE THREAD

**Issue by Rohan487** [781flyingdutchman/background_downloader#119]
**Describe the bug**

Platform: Android 13

I am trying to implement a file download feature in my app, But my wifi files are downloading to 99% and after that, I am getting task error, I attached the logs below, there is no network error or anything like that.  

I already tried both ways of downloading with Foreground enabled and disabled on Android.
I already checked all open and closed issues and tried all fixes nothing works

WHEN I USE MOBILE DATA 5G the download is completed without error.  

**Logs**

![image](https://github.com/781flyingdutchman/background_downloader/assets/72985121/c17be611-75b6-4b0d-8c27-8b4039619cfd)

**Comment by 781flyingdutchman** [comment]
The error message suggests it has something to do with the temp file, but the error message is incomplete. Can you fork the repo (or change directly in your editor) and change one line [here](https://github.com/781flyingdutchman/background_downloader/blob/main/android/src/main/kotlin/com/bbflight/background_downloader/TaskWorker.kt#L610) to: `"Error for taskId ${task.taskId} and $filePath: $e"` so the error message is more complete?

Thanks

**Comment by Rohan487** [comment]
Hey, I followed your instructions, there is a "no such file exception" I Don't know why. I hope it is fixable 

![image](https://github.com/781flyingdutchman/background_downloader/assets/72985121/79dd9c33-1b60-4a77-b31c-d0d60e9a4a99)

**Comment by Rohan487** [comment]
Here are the last progress status logs, you can find additional details from them.  The last progress status was 98%, the test download link is also in the logs.

![image](https://github.com/781flyingdutchman/background_downloader/assets/72985121/9800c75d-e573-4590-878f-8f7b962415b7)

**Comment by 781flyingdutchman** [comment]
Ok, getting closer!  Can you change the same line to `"Error for taskId ${task.taskId} and $filePath: $e\n ${e.stackTrace}"` so we get the trace for where the error occurred, and share the log output?  Also, just to confirm: is there no other error that appears earlier in the stack trace? Thanks.

**Comment by Rohan487** [comment]
> $e\n ${e.stackTrace}

![image](https://github.com/781flyingdutchman/background_downloader/assets/72985121/fc49159f-23ba-4eca-a8d4-768704aa9bdb)


> is there no other error that appears earlier in the stack trace? Thanks.

Not only Task Error, File is downloading fine but at last, it's not saving it to the target dir, maybe some issue with completing file parts or moving file from the cache to the target dir.

**Comment by 781flyingdutchman** [comment]
Yes it seems like the temp file has disappeared before it was copied to the final destination. If you can share the log with the stacktrace we can see if that's indeed true. Still unclear why this would happen, and I'm a bit worried the OS deletes the file for some reason (eg low memory), though that doesn't explain why it works fine over cellular. Anyway, please share the log with stacktrace and we go from there.

**Comment by Rohan487** [comment]
I already shared the logs of stack trace, on above screenshot check there is lIne 

Java.lang.stackTraceElement

after that are normal logs 

#https://github.com/781flyingdutchman/background_downloader/issues/119#issuecomment-1685099621

**Comment by 781flyingdutchman** [comment]
OK, my bad, apparently in Kotlin that doesn't print the actual stackTrace but just that `Java.lang.stackTraceElement` which is not useful. Can you change the line one more time (hopefully!) to: `"Error for taskId ${task.taskId} and $filePath: $e\n ${e.stackTraceToString()}"`

This should tell us exactly where the fileNotFound error happens.

**Comment by Rohan487** [comment]
Got the full stack trace this time.

![image](https://github.com/781flyingdutchman/background_downloader/assets/72985121/e50bf81c-386d-4b87-80be-8877b7ba4971)

**Comment by 781flyingdutchman** [comment]
OK, it looks like the temp file that is used to download the large file is deleted just before it is copied to the final destination. I don't know why - the only thing I can imagine is the OS deleting the file to free up space, so I made one more addition on my side to see how much cache capacity is made available by the system, and if possible I'd like to see what that looks like for yours. In the TaskWorker.kt file line 463 you should see:
`else "${applicationContext.cacheDir}/com.bbflight.background_downloader${Random.nextInt()}"`

After that, insert the following:
```
val storageManager = applicationContext.getSystemService(Context.STORAGE_SERVICE) as StorageManager
val bytesAvailable = storageManager.getCacheQuotaBytes(storageManager.getUuidForPath(applicationContext.cacheDir))
Log.w(TAG, "Cache quota: $bytesAvailable")
```
You'll get a warning that this is only available in API 26 and above, so add above the function (line 440):
```
    @RequiresApi(Build.VERSION_CODES.O)
```

(Android Studio lets you do this by just clicking the action on the squigly red line).

Once you do this and run, you'll see a line in the logs like this:
`TaskWorker              com...background_downloader_example  W  Cache quota: 67108864`
The amount is the quota given to the app by the Android OS (it is not the amount of storage available). I was surprised to see that on my emulator that quota was only about 67MB. If that's true for you, this might explain why this is happening so I am curious to see what values you get. Interestingly, I downloaded the speed.hetzner.de file (100MB) over WiFi just fine despite this cache quota being lower than 100MB, so I still don't quite get what's going on. A few more questions:
1. What Android API version are you using?
2. Are you using the emulator or a real device, and can you test on both?
3. What available memory do you have (on both)?
4. What value for 'Cache quota' do you get from the log (on both)?

Thanks for your patience and help.

**Comment by 781flyingdutchman** [comment]
One more thing, after you've done the logging: in that original line 463, change
`else "${applicationContext.cacheDir}/com.bbflight.background_downloader${Random.nextInt()}"` to `else "${applicationContext.filesDir}/com.bbflight.background_downloader${Random.nextInt()}"` or `else "${applicationContext.dataDir}/com.bbflight.background_downloader${Random.nextInt()}"` and see if that fixes the problem. This changes the place where the temporary file is stored, and should make it immune to deletion by the OS.

**Comment by Rohan487** [comment]
Hey, after changing the temp directory to dataDir as you suggested, it solved all problems of this package. 

The main problem was that OS was deleting the temp file as its in the cache dir, which is solved by changing the temp dir ✅

Second due to the limitations of the cache directory in Android, when I was pausing the file in half the OS was deleting the file and after resuming the file, it started downloading from the beginning, which is now solved ✅ cause in DataDir OS is not deleting the temp file thus the resume of file can be started whenever a user wants 

However I was unable to determine the cache storage of my Android Phone, cause after changing the code I was getting an Unresolved reference: StorageManager error, so I just tried this  #https://github.com/781flyingdutchman/background_downloader/issues/119#issuecomment-1685208485  and it solved major 2 problems of Background downloader package 

Thanks for the help, I will suggest changing the temp dir from the cache to another in next update of background downloader so, OS will not interfere in the download process

**Comment by 781flyingdutchman** [comment]
Ok great and yes it does affect pause as well. Can you let me know what Android version you are running, and the brand & model of your phone?

 I am going to implement this through configuration, because having it default to the dataDir also has drawbacks so I want to give developers a choice. Please check back here on how to do that so you can keep using the published version of the package.

Thanks much for your help in solving a mysterious problem!

**Comment by Rohan487** [comment]
Yes I will update the upcoming new stable version.

My Device details are below , I was testing on same device, Android version is 13 , other details are in image below 

![Screenshot_2023-08-20-22-12-43-35](https://github.com/781flyingdutchman/background_downloader/assets/72985121/0f37a17b-27bd-4da8-b20a-c99ac2312a88)



![Screenshot_2023-08-20-22-12-48-78_fc704e6b13c4fb26bf5e411f75da84f2](https://github.com/781flyingdutchman/background_downloader/assets/72985121/4f944bae-5cb2-4716-8943-6b50ddca32d8)

**Comment by 781flyingdutchman** [comment]
Implemented in V7.9.2

By default, the downloader uses the cacheDir if there is sufficient space, and the dataDir if not. If you fin dyou are still running into issues with this, you can configure the downloader to force use of the dataDir, like this (at the start of your app):
```
FileDownloader().configure(androidConfig: [
      (Config.useCacheDir, Config.never),
    ]).then((result) => debugPrint('Configuration result = $result'));
```

**Comment by 781flyingdutchman** [comment]
Implemented in V7.9.2

By default, the downloader uses the cacheDir if there is sufficient space, and the dataDir if not. If you fin dyou are still running into issues with this, you can configure the downloader to force use of the dataDir, like this (at the start of your app):
```
FileDownloader().configure(androidConfig: [
      (Config.useCacheDir, Config.never),
    ]).then((result) => debugPrint('Configuration result = $result'));
```
